# Lecture
# 1. Importing libraries

In [4]:
# Import libraries
import pandas as pd
import numpy as np
import os

# 2. Importing data

In [32]:
# Create path
path = r'C:\Users\nanh9\Instacart Basket Analysis'
# Orders data
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders.csv'), index_col = False)
# Products data
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)
# Departments data
df_dep = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'departments.csv'), index_col = False) 

# 3. Wrangling data

In [34]:
# dropping column 'eval_set' from df_ords // orders data => df.drop(columns = ['variable'])
df_ords.drop(columns = ['eval_set'])

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0         2539329        1             1          2                  8   
1         2398795        1             2          3                  7   
2          473747        1             3          3                 12   
3         2254736        1             4          4                  7   
4          431534        1             5          4                 15   
...           ...      ...           ...        ...                ...   
3421078   2266710   206209            10          5                 18   
3421079   1854736   206209            11          4                 10   
3421080    626363   206209            12          1                 12   
3421081   2977660   206209            13          1                 12   
3421082    272231   206209            14          6                 14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

In [35]:
# overwriting old dataframe with changes
df_ords = df_ords.drop(columns = ['eval_set'])
# alternative: new dataframe without 'eval_set' => df_ords2 = df_ords.drop(columns = ['eval_set'])

# print all functions within a specific column // df['variable'].value_counts(dropna = False)
df_ords['days_since_prior_order'].value_counts(dropna = False)

days_since_prior_order
30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: count, dtype: int64

In [36]:
# rename column: df.rename(columns = {'old_name' : 'new_name'}, inplace = True)
df_ords.rename(columns = {'order_dow' : 'orders_day_of_week'}, inplace = True)

## Changing data type

In [38]:
df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [39]:
# change data type: astype('datatype') 
df_ords['order_id'].dtype

dtype('int64')

In [40]:
# change data type int64 to str 
df_ords['order_id'] = df_ords['order_id'].astype(str)
df_ords['user_id'] = df_ords['user_id'].astype(str)

In [41]:
df_ords['order_id'].dtype

dtype('O')

## Transposing data

In [43]:
df_dep = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'departments.csv'), index_col = False) 

In [44]:
df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

In [45]:
# change columns to rows with T function
df_dep.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [46]:
# then overwrite the old df_dep with df_dep_t
df_dep_t = df_dep.T

In [47]:
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

### 1) Create a new header for dataframe

In [49]:
# 1)Create a new header // "locate an integer"= iloc[]
new_header = df_dep_t.iloc[0]

In [50]:
new_header

0    department
Name: department_id, dtype: object

In [51]:
# 2) Remove the first row in the dataframe // [1:] from 1 onwards
df_dep_t_new = df_dep_t[1:]

In [52]:
df_dep_t_new

0
1            frozen
2             other
3            bakery
4           produce
5           alcohol
6     international
7         beverages
8              pets
9   dry goods pasta
10             bulk
11    personal care
12     meat seafood
13           pantry
14        breakfast
15     canned goods
16       dairy eggs
17        household
18           babies
19           snacks
20             deli
21          missing

In [53]:
# Add the new header
df_dep_t_new.columns = new_header #set the header row as the df header
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

# 4. Data dictionary

In [55]:
# turn dataframe into dictionary
data_dict = df_dep_t_new.to_dict('index')

In [56]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [57]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [58]:
print(data_dict.get('19')) #use data dictionary to see what department_id = 19 stands for

{'department': 'snacks'}


### Subsetting

In [60]:
# use criteria to filter: subset of 'df_prods' that only contains data from 'snacks' department
df_snacks = df_prods[df_prods['department_id']==19]

In [61]:
# Indexing => syntax: object['element']  
df_prods['department_id']==19

0         True
1        False
2        False
3        False
4        False
         ...  
49688    False
49689    False
49690    False
49691    False
49692    False
Name: department_id, Length: 49693, dtype: bool

In [62]:
# enclosing df_prods['department_id']==19 inside df_prods[] => shows list ('snacks') of dataframe that only contains department_id = 19
df_prods[df_prods['department_id']==19]

product_id                                   product_name  aisle_id  \
0               1                     Chocolate Sandwich Cookies        61   
15             16                  Mint Chocolate Flavored Syrup       103   
24             25        Salted Caramel Lean Protein & Fiber Bar         3   
31             32                  Nacho Cheese White Bean Chips       107   
40             41    Organic Sourdough Einkorn Crackers Rosemary        78   
...           ...                                            ...       ...   
49666       49662                   Bacon Cheddar Pretzel Pieces       107   
49669       49665  Super Dark Coconut Ash & Banana Chocolate Bar        45   
49670       49666                  Ginger Snaps Snacking Cookies        61   
49675       49671                           Milk Chocolate Drops        45   
49683       49679                        Famous Chocolate Wafers        61   

       department_id  prices  
0                 19     5.8  
15                19     5.2  
24                19     1.9  
31                19     4.9  
40                19     6.5  
...              ...     ...  
49666             19     3.6  
49669             19     6.9  
49670             19     5.2  
49675             19     3.0  
49683             19     6.0  

[6264 rows x 5 columns]

In [63]:
df_snacks = df_prods[df_prods['department_id']==19]
df_snacks.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

In [64]:
# alternative way #1: loc function => locate column
df_snacks_3 = df_prods.loc[df_prods['department_id']==19]

In [65]:
# alternative way #2: use loc function to look into a list with more than one value: isin([19])
df_snacks_3 = df_prods.loc[df_prods['department_id'].isin([19])]

# 5. Exporting dataframes

In [67]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_wrangled.csv'))

---

# Assignment

### 2) Find another identifier variable in the df_ords dataframe that doesn't need to be included in your analysis as a numeric variable and change it to a suitable format.

In [71]:
# user_id = indentifier & changing it to string
df_ords['user_id'].dtype

dtype('O')

In [72]:
df_ords['user_id'] = df_ords['user_id'].astype(str)

In [73]:
df_ords.describe()

order_number  orders_day_of_week  order_hour_of_day  \
count  3.421083e+06        3.421083e+06       3.421083e+06   
mean   1.715486e+01        2.776219e+00       1.345202e+01   
std    1.773316e+01        2.046829e+00       4.226088e+00   
min    1.000000e+00        0.000000e+00       0.000000e+00   
25%    5.000000e+00        1.000000e+00       1.000000e+01   
50%    1.100000e+01        3.000000e+00       1.300000e+01   
75%    2.300000e+01        5.000000e+00       1.600000e+01   
max    1.000000e+02        6.000000e+00       2.300000e+01   

       days_since_prior_order  
count            3.214874e+06  
mean             1.111484e+01  
std              9.206737e+00  
min              0.000000e+00  
25%              4.000000e+00  
50%              7.000000e+00  
75%              1.500000e+01  
max              3.000000e+01

-> The identifiers order_id & user_id are now removed.

### 3) Look for a variable in your df_ords dataframe with an unintuitive name and change its name without overwriting the dataframe.

'days_since_prior_order' seems to look odd. So changing it to 'days_since_last_order' should be the right move.

In [77]:
df_ords.rename(columns = {'days_since_prior_order' : 'days_since_last_order'}, inplace = True)

In [78]:
# check changes
df_ords.describe()

order_number  orders_day_of_week  order_hour_of_day  \
count  3.421083e+06        3.421083e+06       3.421083e+06   
mean   1.715486e+01        2.776219e+00       1.345202e+01   
std    1.773316e+01        2.046829e+00       4.226088e+00   
min    1.000000e+00        0.000000e+00       0.000000e+00   
25%    5.000000e+00        1.000000e+00       1.000000e+01   
50%    1.100000e+01        3.000000e+00       1.300000e+01   
75%    2.300000e+01        5.000000e+00       1.600000e+01   
max    1.000000e+02        6.000000e+00       2.300000e+01   

       days_since_last_order  
count           3.214874e+06  
mean            1.111484e+01  
std             9.206737e+00  
min             0.000000e+00  
25%             4.000000e+00  
50%             7.000000e+00  
75%             1.500000e+01  
max             3.000000e+01

### 4. Your client wants to know what the busiest hour is for placing orders. Find the frequency of the corresponding variable and share your findings.

In [80]:
df_ords['order_hour_of_day'].value_counts(dropna = False)

order_hour_of_day
10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: count, dtype: int64

=> The busiest hour for orders is from 10am to 11am!

### 5. Determine the meaning behind a value of 4 in the "department_id" column within the df_prods dataframe using a data dictionary.

In [83]:
# check data dictionary for department_id
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [84]:
print(data_dict.get('4'))

{'department': 'produce'}


=> 'department_id' = 4 stands for 'produce'

### 6. The sales team in your client's organization wants to know more about breakfast item sales. Create a subset containing only the required information.

In [87]:
#creating subset: df_breakfast
df_breakfast = df_prods[df_prods['department_id']==14]

In [88]:
df_breakfast

product_id                                       product_name  \
27             28                                  Wheat Chex Cereal   
33             34                                                NaN   
67             68                            Pancake Mix, Buttermilk   
89             90                                       Smorz Cereal   
210           211   Gluten Free Organic Cereal Coconut Maple Vanilla   
...           ...                                                ...   
49330       49326                            Cereal Variety Fun Pack   
49395       49391            Light and Fluffy Buttermilk Pancake Mix   
49547       49543                          Chocolate Cheerios Cereal   
49637       49633               Shake 'N Pour Buttermilk Pancake Mix   
49667       49663  Ultra Protein Power Crunch Peanut Butter N' Ho...   

       aisle_id  department_id  prices  
27          121             14    10.1  
33          121             14    12.2  
67          130             14    13.7  
89          121             14     3.9  
210         130             14     3.6  
...         ...            ...     ...  
49330       121             14     9.1  
49395       130             14     2.0  
49547       121             14    10.8  
49637       130             14    14.2  
49667        57             14    10.2  

[1116 rows x 5 columns]

### 7. They’d also like to see details about products that customers might use to throw dinner parties. Your task is to find all observations from the entire dataframe that include items from the following departments: alcohol, deli, beverages, and meat/seafood. You’ll need to present this subset to your client.

data_dict tells us the value of 5, 20, 7, 12 

In [91]:
df_dinner_party = df_prods.loc[df_prods['department_id'].isin([5, 7, 12, 20])]

In [92]:
df_dinner_party

product_id                                    product_name  aisle_id  \
2               3            Robust Golden Unsweetened Oolong Tea        94   
6               7                  Pure Coconut Water With Orange        98   
9              10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10             11                               Peach Mango Juice        31   
16             17                               Rendered Duck Fat        35   
...           ...                                             ...       ...   
49676       49672                          Cafe Mocha K-Cup Packs        26   
49679       49675             Cinnamon Dolce Keurig Brewed K Cups        26   
49680       49676                          Ultra Red Energy Drink        64   
49686       49682                              California Limeade        98   
49688       49684       Vodka, Triple Distilled, Twist of Vanilla       124   

       department_id  prices  
2                  7     4.5  
6                  7     4.4  
9                  7     8.4  
10                 7     2.8  
16                12    17.1  
...              ...     ...  
49676              7     6.5  
49679              7    14.0  
49680              7    14.5  
49686              7     4.3  
49688              5     5.3  

[7650 rows x 5 columns]

### 8. It's important that you keep track of total counts in your dataframes. How many rows does the last dataframe you created have?

In [94]:
df_dinner_party.shape

(7650, 5)

=> The last dataframe df_dinner_party has 7650 rows x 5 columns.

### 9. Someone from the data engineers team in Instacart thinks they've spotted something strange about the customer with a "user_id" of "1." Extract all the information you can about the user.

In [97]:
df_ords_user_1 = df_ords[df_ords['user_id']=='1']

In [98]:
df_ords_user_1

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329       1             1                   2                  8   
1   2398795       1             2                   3                  7   
2    473747       1             3                   3                 12   
3   2254736       1             4                   4                  7   
4    431534       1             5                   4                 15   
5   3367565       1             6                   2                  7   
6    550135       1             7                   1                  9   
7   3108588       1             8                   1                 14   
8   2295261       1             9                   1                 16   
9   2550362       1            10                   4                  8   
10  1187899       1            11                   4                  8   

    days_since_last_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0  
5                    19.0  
6                    20.0  
7                    14.0  
8                     0.0  
9                    30.0  
10                   14.0

### 10. You also need to provide some details about this user's behavior. What basic stats can you provide based on the information you have?

In [100]:
df_ords_user_1.describe()

order_number  orders_day_of_week  order_hour_of_day  \
count     11.000000           11.000000          11.000000   
mean       6.000000            2.636364          10.090909   
std        3.316625            1.286291           3.477198   
min        1.000000            1.000000           7.000000   
25%        3.500000            1.500000           7.500000   
50%        6.000000            3.000000           8.000000   
75%        8.500000            4.000000          13.000000   
max       11.000000            4.000000          16.000000   

       days_since_last_order  
count              10.000000  
mean               19.000000  
std                 9.030811  
min                 0.000000  
25%                14.250000  
50%                19.500000  
75%                26.250000  
max                30.000000

=> User 1 orders 4 days a week. I assume 0=Sunday, which leads me to the conclusion that user_1 orders from Sunday to Wednesday. User_1 made most of the orders between 7 and 9am. The average order frequency is 19 days between orders. If we ignore the null value of 'days_since_last_order' user_1 doesn't wait longer than 30 days for the next order.

### 11. Check the organization and structure of your notebook. Be sure to include section headings and code comments.

# 6. Export data

In [104]:
# export orders dataframe
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_wrangled.csv'))

In [105]:
#export 
df_dep_t_new.to_csv(os.path.join(path, '02 Data','Prepared Data', 'departments_wrangled.csv'))